include/pcmki/pcmki_output.h

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2019 Chris Lumens <clumens@redhat.com>
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#ifndef CRM_OUTPUT__H
+#  define CRM_OUTPUT__H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#  include <stdio.h>
+#  include <libxml/tree.h>
+
+#  include <glib.h>
+#  include <crm/stonith-ng.h>
+
+#  define PCMK_API_VERSION "0.1"
+
+typedef struct pcmk__output_fmt_s pcmk__output_fmt_t;
+typedef struct pcmk__output_s pcmk__output_t;
+
+typedef enum {
+    pcmk_output_text,
+    pcmk_output_xml
+} pcmk__output_type;
+
+struct pcmk__output_fmt_s {
+    /*!
+     * \brief Take whatever actions are necessary to prepare out for use.  This is
+     *        called by pcmk__output_new.  End users should not need to call this.
+     *
+     * \note Must be called before any other functions in this structure.
+     *
+     * \param out The output functions structure.
+     *
+     * \return true on success, false on error.
+     */
+    gboolean (*init) (pcmk__output_t *out);
+
+    /*!
+     * \brief Take whatever actions are necessary to end formatted output.  This
+     *        could include flushing output to a file, but does not include freeing
+     *        anything.
+     *
+     * \note No other functions in this structure may be called after this one.
+     *
+     * \param out The output functions structure.
+     */
+    void (*finish) (pcmk__output_t *out);
+
+    /*!
+     * \brief Format the output of executed command.
+     *
+     * \param out         The output functions structure.
+     * \param output_type The source of this output - stdout or stderr.
+     * \param buf         The raw command output to be formatted.
+     */
+    void (*command_output) (pcmk__output_t *out, const char *output_type, const char *buf);
+
+    /*!
+     * \brief Format an information message that should be shown to
+     *        to the user.  Not all formatters will do this.
+     *
+     * \param out    The output functions structure.
+     * \param format The printf-style format string to use in output.
+     */
+    void (*information) (pcmk__output_t *out, const char *format, ...)
+         G_GNUC_PRINTF(2, 3);
```

Comment:
I'm going to change this function (and end_list_with_summary below) to get rid of the format string argument potential security problem.  I've got an idea for how to do it, and I'll take a look in the morning.